### ISSUE TITLE
'"x" is not a known member of None' after a raise inside preceding function call

### ISSUE THREAD

**Issue by K1DV5** [microsoft/pyright#5795]
**Describe the bug**
As shown in the code below, while there is a check inside the preceding function call if the value is None that raises an error, which should block the execution from getting there with None, Pyright reports it as a violation. Interestingly, If we replace the check function call with its body, Pyright is ok with that.

**Code or Screenshots**
![image](https://github.com/microsoft/pyright/assets/40162008/e19c06a0-ac85-479f-9a06-0dd6b1d3c94e)
```python
def check(r: int | None):
    if r is None:
        raise ValueError()

def foo(r: int | None):
    check(r)
    return r.to_bytes()
```

**VS Code extension or command-line**
I'm using pyright on Neovim 0.9

**Comment by tmke8** [comment]
This also doesn't seem to work:

```python
from typing import NoReturn, overload

@overload
def check(r: None) -> NoReturn: ...
@overload
def check(r: int) -> None: ...
def check(r: int | None) -> None:
    if r is None:
        raise ValueError()

def foo(r: int | None):
    check(r)
    return r.to_bytes()  # still error
```

**Comment by erictraut** [comment]
Pyright is working as designed here. Static type checkers don't execute the code, nor do they perform global (cross-function) analysis. Type narrowing is performed only locally within a function, not across functions.

Type systems in some languages (such as TypeScript) support the feature you're looking for. It's referred to as a "type assert". I have proposed to add a similar capability to Python. You can read about it [here](https://github.com/python/typing/discussions/1013#discussioncomment-1966238). There has recently been some additional interest expressed in this idea, so it's possible that someone will write a PEP and try to get it accepted. If you're interested in this feature, you could lend your support by responding to that thread.

In the meantime, you can leverage the type narrowing support that is already supported in pyright. For complete details, please refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing).
